std/algorithm/comparison.d

```diff
@@ -530,6 +529,7 @@ Returns:
 
 */
 auto clamp(T1, T2, T3)(T1 val, T2 lower, T3 upper)
+if (is(typeof(max(lower, min(upper, val)))))
```

Comment:
eh. This is kind of annoying. You are repeating the entire function in the constraints, which means that you will get a matching error with the provided code instead of a more specific error. I would recommend against this.

In this case, it's simple, but in the general case, I don't like repeating the entire function in the constraints, unless there is a need to do this (e.g. for overloading). Later, if you change one but not the other, really confusing things happen on corner cases.

Reply:
The error message looks OK, and this example is somewhat saved by the fact that the current error message looks laughably confusing:

```D
import std.algorithm;
auto x = clamp("2", 1, 3);
```

```
/dlang/dmd/linux/bin64/../../src/phobos/std/algorithm/comparison.d(1634): Error: no property max for type string, perhaps import std.algorithm; is needed?
/dlang/dmd/linux/bin64/../../src/phobos/std/algorithm/comparison.d(1664): Error: template instance std.algorithm.comparison.MinType!(int, string) error instantiating
/dlang/dmd/linux/bin64/../../src/phobos/std/algorithm/comparison.d(540):        instantiated from here: min!(int, string)
onlineapp.d(11):        instantiated from here: clamp!(string, int, int)
```

But this is actually a problem with `MinType`, which makes some incorrect assumptions (if T[0] has a min property, then all T must have min/max properties). A better error message down in `min` would be useful not just for this case.

My preference is to leave the constraint off, and fix the error message from min/max to be useful. I don't think even for one-liners we should be repeating the entire code in the constraint.

An exception would be for cases where constraints are needed for overloading.

Reply:
MinType will be eliminated by another PR. Not sure where this would leave us.

Phobos functions have constraints that prevent them from accepting arbitrary arguments, to then fail in their own code in arbitrary ways. `clamp` was an outlier, and this PR fixes that.

Reply:
Yeah, actually, the error getting pushed down doesn't help. Because min or max has a similar error problem of gagging actual errors that might be useful.